File: neuralnetwork/Network.js
class Network {
    inputs;
    outputs;
    layers = [];
    depths;

    constructor(inputs, outputs) {
        this.inputs = inputs;
        this.outputs = outputs;
    }

    generate(depths) {
        this.depths = depths
        for (var i = 0; i < depths.length + 1; i++) {
            this.layers[i] = []
            for (var j = 0; j < [this.inputs, ...depths, this.outputs][i]; j++) {
                this.layers[i][j] = []
                for (var z = 0; z < [this.inputs, ...depths, this.outputs][i + 1]; z++) {
                    this.layers[i][j][z] = (1 - 2 * Math.random()) * 10
                }
            }
        }

    }

    calculate(input) {
        var nextLayer = [];
        for (var i = 0; i < this.layers.length; i++) {
            for (var j = 0; j < this.layers[i][0].length; j++) {
                nextLayer[j] = 0;
                for (var z = 0; z < input.length; z++) {
                    nextLayer[j] += input[z] * this.layers[i][z][j]
                }
            }
            input = []
            for (let j = 0; j < nextLayer.length; j++) {
                input[j] = nextLayer[j]
            }
            nextLayer = []
        }
        return input;
    }

    calculateGPU(input) {
        const gpu = new GPU();
        const multiplyMatrix = gpu.createKernel(function (a, b, size) {
            let sum = 0;
            for (let i = 0; i < size; i++) {
                sum += a[this.thread.y][i] * b[i][this.thread.x];
            }
            return sum;
        }, {
            dynamicOutput: true,
            dynamicArguments: true
        })

        for (let i = 0; i < this.layers.length; i++) {
            multiplyMatrix.setOutput([this.layers[i][0].length, 1])
            input = multiplyMatrix(input, this.layers[i], this.layers[i].length)
        }
        return input;
    }

    mutate(percentage = 0.8, offset = 0.1) {
        var mutated = new Network(this.inputs, this.outputs)

        for (var i = 0; i < this.layers.length; i++) {

            mutated.layers[i] = []
            for (var j = 0; j < this.layers[i].length; j++) {
                mutated.layers[i][j] = []
                for (var z = 0; z < this.layers[i][j].length; z++) {
                    if (Math.random() < percentage) {
                        mutated.layers[i][j][z] = this.layers[i][j][z] + (1 - 2 * Math.random()) * offset

                    } else {
                        mutated.layers[i][j][z] = this.layers[i][j][z]
                    }
                }

            }
        }
        return mutated;
    }



    get getLayers() {
        return this.layers
    }
}

module.exports = {Network}

File: cpp/NeuralNetwork/source/Network.cpp
#include "../headers/Network.h"


Network::Network(int inputs_t, int outputs_t, void(*activation_t)(double&)){
    inputs = inputs_t;
    outputs = outputs_t;
    activation = activation_t;
}

Network::Network(){
};

Network::~Network(){
}

void Network::generate(int depths_t[], int size){

    layerCount = size - 1;
    depths = (int*)malloc(sizeof(int)*size);
    int cellCount = 0;

    for(int i = 0; i < size-1; i++){
        depths[i] = depths_t[i];
        cellCount += depths_t[i] * depths_t[i+1];
    }
    depths[size-1] = depths_t[size-1];
    layers = (float*)malloc(sizeof(float)*cellCount);
    for(int i = 0; i < cellCount; i++){
        layers[i] = (1 - (rand() % 100) / 50)*1;
        
    }
}

double* Network::calculate(double* input){
    int maxDepth = 0;

    for(int i = 0; i < layerCount + 1; i++){
        if(depths[i] > maxDepth){maxDepth = depths[i];}
    }

    double* inputLayer = (double*)malloc(sizeof(double) * maxDepth);
    double* outputLayer = (double*)malloc(sizeof(double) * maxDepth);

    for(int i = 0; i<inputs; i++){
        inputLayer[i] = input[i];
    }

    int offset = 0;
    double* t;

    for(int i = 0; i < layerCount; i++){
        if(i != 0){
            offset += depths[i-1] * depths[i];
        }
        for(int j = 0; j < depths[i+1]; j++){
            outputLayer[j] = 0;
            for(int z = 0; z < depths[i]; z++){
                outputLayer[j] += inputLayer[z] * layers[offset + z*depths[i+1] + j];
            }
            activation(outputLayer[j]);
        }
        t = inputLayer;
        inputLayer = outputLayer;
        outputLayer = t;
    }
    free(outputLayer);
    return inputLayer;
}

Network Network::mutate(float percentage, float offset){
    
    Network mutated = Network(inputs, outputs, activation);
    mutated.layerCount = layerCount;
    mutated.depths = (int*)malloc(sizeof(int)*(layerCount+1));
    int cellCount = 0;

    for(int i = 0; i < layerCount; i++){
        mutated.depths[i] = depths[i];
        cellCount += depths[i] * depths[i+1];
    }

    mutated.depths[layerCount] = depths[layerCount];
    mutated.layers = (float*)malloc(sizeof(float)*cellCount);

    for(int i = 0; i < cellCount; i++){
        if((rand() % 100) / 100 < percentage){
            mutated.layers[i] = layers[i] + (1 -float(rand() % 100) / 50) * offset;
        }else{
            mutated.layers[i] = layers[i];
        }
        
    }
    return mutated;
}

int Network::getComplexity(){
    int count = 0;
    for(int i = 0; i < layerCount; i++){
        count += depths[i] * depths[i+1];
    }
    return count;
}
